### ISSUE TITLE
False positive for -Wdangling-initializer-list

### ISSUE THREAD

**Issue by CaseyCarter** [llvm/llvm-project#24555]
|  |  |
| --- | --- |
| Bugzilla Link | [24181](https://llvm.org/bz24181) |
| Version | trunk |
| OS | Linux |
| Attachments | [Test case](https://user-images.githubusercontent.com/456873/143751737-09c0a580-133a-40dd-b0bc-c406080066e0.gz) |
| CC | @DougGregor |

## Extended Description 

Clang's dangling-initializer-list diagnostic is incorrectly emitted twice for this program:

#include <initializer_list>

int main() {
    new std::initializer_list<int>[2]{};
    new std::initializer_list<int>{};
}

All instances of std::initializer_list in this program are value-initialized, and hence not backed by any array.

**Comment by wheatman** [comment]
This was fixed between 6.0.1 and 7.0.0
https://godbolt.org/z/sbPTE817r
code
```
#include <initializer_list>

int main() {
    new std::initializer_list<int>[2]{};
    new std::initializer_list<int>{};
}
```

old warning

```
<source>:4:9: warning: array backing the initializer list will be destroyed at the end of the full-expression [-Wdangling-initializer-list]
    new std::initializer_list<int>[2]{};
        ^                            ~~
<source>:5:9: warning: array backing the initializer list will be destroyed at the end of the full-expression [-Wdangling-initializer-list]
    new std::initializer_list<int>{};
        ^                         ~~
2 warnings generated.
Compiler returned: 0
```
